projects/igniteui-angular/src/lib/grids/grid-navigation.service.ts

```diff
@@ -257,17 +259,31 @@ export class IgxGridNavigationService {
         }
     }
 
+    private findFirstDataRowIndex() {
+        const dv = this.grid.dataView;
+        return dv.findIndex(rec => !this.grid.isGroupByRecord(rec) && !this.grid.isDetailRecord(rec));
+    }
+
+    private findLastDataRowIndex() {
```

Comment:
While this function would not be called by a continuous key press making its performance less important, don't you think it would be a lot better implementing it like this:

```typescript
private findLastDataRowIndex() {
    let i = this.grid.dataView.length;
    while (i--) {
        const rec = this.grid.dataView[i];
        if (!this.grid.isGroupByRecord(rec) && !this.grid.isDetailRecord(rec)) {
             return i;
        }
    }
}
```
In this way we avoid the reverse call and the additional index search in the original array, which in the most common case will have to iterate it again fully.

Unfortunately, ES still doesn't define reverse find functions but you could also create one if you want this particular code to be clean.